File: internal/handlers/user_auth.go
package handlers

import (
	"context"
	"log"
	"net/http"
	"server/internal/utils"
	"time"

	"github.com/labstack/echo/v4"
)

// Signin user
func (h *Handler) SignInAuth(c echo.Context) error {
	state, err := utils.GenerateRandomstring()
	if err != nil {
		log.Println(err)
		return c.String(500, "Internal server error")
	}

	// Set state cookie
	c.SetCookie(&http.Cookie{
		Name:     "state",
		Value:    state,
		HttpOnly: true,
		Path:     "/",
		Secure:   true,
		Expires:  time.Now().Add(5 * time.Minute),
	})

	// Add oauth2.AccessTypeOffline to get a refresh token
	return c.Redirect(307, h.SignInAuthConfig.AuthCodeURL(state))
}

// Sign out user
func (h *Handler) Signout(c echo.Context) error {
	ctx := context.Background()

	cookie, err := c.Request().Cookie("session_id")
	if err != nil {
		log.Println(err)
	}

	if cookie.Value == "" {
		log.Println("session id cookie is empty")
	}

	sessionID := cookie.Value

	// TODO: Delete sign in token

	// Delete the user id from the redis cache
	_, rErr := h.RedisClient.Del(ctx, sessionID).Result()
	if rErr != nil {
		log.Println(rErr)
	}

	utils.ClearSessionIDCookie(c)

	return c.Redirect(307, "/")
}

File: routes/user_auth.py
from fastapi import APIRouter, Request
from fastapi.responses import RedirectResponse
from datetime import datetime, timedelta, timezone
from utils.auth_utils import sign_in_auth_config

router = APIRouter()

@router.get("/sign-in")
async def sign_in(request: Request):
    auth_config = sign_in_auth_config()
    auth_url, state = auth_config.authorization_url(
       #  access_type="offline",
       # include_granted_scopes="true",
       # prompt="consent"
    )

    response = RedirectResponse(url=auth_url, status_code=302)
    expires = datetime.now(timezone.utc) + timedelta(minutes=5)
    response.set_cookie(
        key="state",
        value=state,
        expires=expires,
        path="/",
        secure=True,
        httponly=True,
        samesite="lax"
    )

    return response


@router.get("/sign-out")
async def sign_out(request: Request):
    # Delete session id from redis
    # Clear session id cookie
    # Redirect to home page
    pass
